File: android/app/src/main/java/com/example/jamboree/games/president/PresidentServer.kt
package com.example.jamboree.games.president

import android.content.Context
import com.example.jamboree.join.LoopbackGuest
import com.example.jamboree.social.GuestId
import com.example.jamboree.social.HostServer
import com.example.jamboree.stats.SeriesScore
import com.example.jamboree.stats.StatsStore
import com.example.jamboree.tutorials.GameTutorials
import org.json.JSONArray
import org.json.JSONObject

/** Wraps [HostServer] with President (Scum/Asshole) routing. */
class PresidentServer(context: Context, val hostName: String = "Host") {
    var engine = PresidentEngine(); private set
    private val server = HostServer(html = HostServer.htmlAsset(context, "president_browser.html"), ctx = context)
    private val appCtx = context.applicationContext
    private val guestToPlayer = HashMap<GuestId, String>()
    private val playerToGuest = HashMap<String, GuestId>()
    var onStateChange: (() -> Unit)? = null
    private var statRecorded = false
    private val series = SeriesScore()
    // President plays multiple rounds per match (next_round), so the series
    // counts each round's president — gated separately from the once-per-match
    // stat recording.
    private var seriesRecorded = false

    init { server.onMessage = ::onMessage; server.onLeave = ::onLeave }
    companion object { const val HOST_ID = "host" }

    fun start(): String? {
        engine.addPlayer(HOST_ID, hostName, isHost = true)
        val u = server.startServer()
        val local = server.attachLocalGuest()
        guestToPlayer[local] = HOST_ID
        playerToGuest[HOST_ID] = local
        emit()
        return u
    }

    fun makeLoopback(): LoopbackGuest = LoopbackGuest(server)
    fun stop() { server.stopServer(); resetState() }

    /** Clear all per-session state so the next time the host starts
     *  hosting they get a fresh screen — empty lobby, tutorial vote
     *  available again. */
    private fun resetState() {
        engine = PresidentEngine()
        guestToPlayer.clear(); playerToGuest.clear()
        series.reset(); seriesRecorded = false
        statRecorded = false
        emit()
    }
    val guestCount: Int get() = server.guestCount

    fun hostSetOptions(o: PresOptions) {
        engine.setOptions(o); broadcastOptions(); emit()
    }
    fun hostStart() {
        seriesRecorded = false
        engine.start(); broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
    }
    fun hostNextRound() {
        seriesRecorded = false
        engine.startNextRound(); broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
    }
    fun hostNewGame() {
        engine.reset(); statRecorded = false
        broadcast(JSONObject().put("type", "reset")); broadcastLobby(); broadcastTutorialState()
        if (!series.isEmpty) broadcastSeries()
        emit()
    }
    fun hostCallTutorialVote() = openTutorialVote()
    fun hostTutorialVote(yes: Boolean) = submitTutorialVote(HOST_ID, yes)
    fun hostDismissTutorial() { engine.tutorialVote.markShown(); broadcastTutorialState(); emit() }

    private fun onMessage(guest: GuestId, raw: String) {
        val j = runCatching { JSONObject(raw) }.getOrNull() ?: return
        val pid = guestToPlayer[guest]
        when (j.optString("type")) {
            "join" -> handleJoin(guest, j)
            "play" -> pid?.let { applyPlay(it, j) }
            "pass" -> pid?.let { applyPass(it) }
            "swap" -> pid?.let { applySwap(it, j) }
            "next_round" -> if (engine.phase == PresidentPhase.GAME_OVER) hostNextRound()
            "set_options" -> Unit
            "call_tutorial_vote" -> openTutorialVote()
            "tutorial_vote" -> pid?.let { submitTutorialVote(it, j.getBoolean("yes")) }
        }
    }

    private fun onLeave(guest: GuestId) {
        val pid = guestToPlayer.remove(guest) ?: return
        playerToGuest.remove(pid)
        if (engine.phase == PresidentPhase.LOBBY) {
            engine.removePlayer(pid); engine.tutorialVote.removeVoter(pid); broadcastLobby()
            if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) broadcastTutorialState()
        }
        emit()
    }

    private fun handleJoin(guest: GuestId, j: JSONObject) {
        if (engine.phase != PresidentPhase.LOBBY) {
            send(guest, JSONObject().put("type", "error").put("message", "Game already in progress")); return
        }
        val name = j.optString("name").trim(); if (name.isEmpty()) return
        val pid = "g${guestToPlayer.size + 1}"
        engine.addPlayer(pid, name)
        guestToPlayer[guest] = pid; playerToGuest[pid] = guest
        send(guest, JSONObject().put("type", "welcome").put("yourId", pid).put("yourName", name).put("game", "president"))
        broadcastLobby(); broadcastOptions(); broadcastTutorialState()
        if (!series.isEmpty) broadcastSeries()
        emit()
    }

    private fun applyPlay(pid: String, j: JSONObject) {
        val cards = parseCards(j.optJSONArray("cards")) ?: return
        val err = engine.play(pid, cards)
        if (err == null) {
            broadcastState(); sendHandsPrivately()
            if (engine.phase == PresidentPhase.GAME_OVER) broadcastOver()
            emit()
        } else playerToGuest[pid]?.let { send(it, JSONObject().put("type", "error").put("message", err)) }
    }

    private fun applyPass(pid: String) {
        val err = engine.pass(pid)
        if (err == null) { broadcastState(); emit() }
        else playerToGuest[pid]?.let { send(it, JSONObject().put("type", "error").put("message", err)) }
    }

    private fun applySwap(pid: String, j: JSONObject) {
        val cards = parseCards(j.optJSONArray("cards")) ?: emptyList()
        val err = engine.submitSwap(pid, cards)
        if (err == null) {
            broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
        } else playerToGuest[pid]?.let { send(it, JSONObject().put("type", "error").put("message", err)) }
    }

    private fun parseCards(arr: JSONArray?): List<PresCard>? {
        if (arr == null) return null
        val out = ArrayList<PresCard>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            val suit = PresSuit.entries.firstOrNull { it.name.equals(o.getString("suit"), true) } ?: return null
            out.add(PresCard(suit, o.getInt("rank")))
        }
        return out
    }

    private fun broadcastOptions() {
        broadcast(JSONObject().put("type", "options")
            .put("allowHouseRules", engine.options.allowHouseRules)
            .put("revolution", engine.options.revolution))
    }

    private fun broadcastLobby() {
        val arr = JSONArray()
        for (p in engine.players.values) arr.put(JSONObject().put("id", p.id).put("name", p.name).put("isHost", p.isHost))
        broadcast(JSONObject().put("type", "lobby").put("players", arr))
    }

    private fun broadcastState() {
        val playersArr = JSONArray()
        for (p in engine.players.values) {
            playersArr.put(JSONObject()
                .put("id", p.id).put("name", p.name)
                .put("handCount", p.hand.size)
                .put("rank", p.rank.wireValue())
                .put("finished", p.finished)
                .put("finishOrder", p.finishOrder))
        }
        val phase = when (engine.phase) {
            PresidentPhase.LOBBY -> "lobby"
            PresidentPhase.SWAPPING -> "swapping"
            PresidentPhase.PLAYING -> "playing"
            PresidentPhase.GAME_OVER -> "gameOver"
        }
        val payload = JSONObject()
            .put("type", "state").put("phase", phase)
            .put("currentId", engine.current?.id ?: JSONObject.NULL)
            .put("roundNumber", engine.roundNumber)
            .put("lastEvent", engine.lastEvent ?: "")
            .put("players", playersArr)
            .put("seating", JSONArray(engine.seatingSnapshot))
            .put("passedThisTrick", JSONArray(engine.passedThisTrick.toList()))
        engine.trick?.let { t ->
            val obj = comboJson(t.combo).put("topPower", t.topPower).put("leaderId", t.leaderId)
            payload.put("trick", obj)
        }
        engine.lastPlay?.let { lp ->
            val cards = JSONArray()
            for (c in lp.cards) cards.put(JSONObject().put("suit", c.suit.name.lowercase()).put("rank", c.rank))
            payload.put("lastPlay", JSONObject()
                .put("playerId", lp.playerId)
                .put("cards", cards)
                .put("combo", comboJson(lp.combo)))
        }
        broadcast(payload)
    }

    private fun comboJson(c: PresCombo): JSONObject = when (c) {
        is PresCombo.Single -> JSONObject().put("kind", "single").put("length", 1)
        is PresCombo.Pair -> JSONObject().put("kind", "pair").put("length", 2)
        is PresCombo.Triple -> JSONObject().put("kind", "triple").put("length", 3)
        is PresCombo.Quad -> JSONObject().put("kind", "quad").put("length", 4)
        is PresCombo.RunOfPairs -> JSONObject().put("kind", "runOfPairs").put("length", c.length * 2)
    }

    private fun sendHandsPrivately() {
        for (p in engine.players.values) {
            val g = playerToGuest[p.id] ?: continue
            val cards = JSONArray()
            for (c in p.hand) cards.put(JSONObject().put("suit", c.suit.name.lowercase()).put("rank", c.rank))
            send(g, JSONObject().put("type", "hand").put("cards", cards))
        }
    }

    private fun broadcastSwapPrompts() {
        for (p in engine.players.values) {
            val g = playerToGuest[p.id] ?: continue
            val arr = JSONArray()
            for (sw in engine.pendingSwaps) {
                if (sw.cards != null || sw.fromId != p.id) continue
                arr.put(JSONObject()
                    .put("toId", sw.toId)
                    .put("toName", engine.players[sw.toId]?.name ?: "")
                    .put("count", sw.count)
                    .put("giverChooses", sw.giverChooses))
            }
            send(g, JSONObject().put("type", "swap_prompts").put("prompts", arr))
        }
    }

    private fun broadcastOver() {
        if (!statRecorded) {
            statRecorded = true
            val names = ArrayList<String>()
            for (pid in engine.finishOrder) engine.players[pid]?.let { names.add(it.name) }
            StatsStore.record(appCtx, "president", names, "win")
        }
        if (!seriesRecorded) {
            seriesRecorded = true
            val presidentName = engine.finishOrder.firstOrNull()?.let { engine.players[it]?.name }
            if (presidentName != null) { series.record(presidentName); broadcastSeries() }
        }
        val rankings = JSONArray()
        for (pid in engine.finishOrder) {
            val p = engine.players[pid] ?: continue
            rankings.put(JSONObject()
                .put("id", p.id).put("name", p.name)
                .put("rank", p.rank.wireValue())
                .put("finishOrder", p.finishOrder))
        }
        broadcast(JSONObject().put("type", "over").put("rankings", rankings))
    }

    private fun openTutorialVote() {
        if (engine.phase != PresidentPhase.LOBBY) return
        if (engine.tutorialVote.isOpen || engine.tutorialVote.tutorialShown) return
        engine.tutorialVote.open(engine.players.keys); broadcastTutorialState(); emit()
    }
    private fun submitTutorialVote(voterId: String, yes: Boolean) {
        if (!engine.tutorialVote.isOpen) return
        engine.tutorialVote.submit(voterId, yes); broadcastTutorialState(); emit()
    }
    private fun broadcastTutorialState() {
        val v = engine.tutorialVote
        val p = JSONObject().put("type", "tutorial_vote_state")
            .put("isOpen", v.isOpen).put("yesCount", v.yesCount).put("noCount", v.noCount)
            .put("eligibleCount", v.eligibleCount).put("result", v.result ?: JSONObject.NULL)
            .put("tutorialShown", v.tutorialShown)
        if (v.result == true && !v.tutorialShown) {
            p.put("title", GameTutorials.president.title)
            p.put("sections", JSONArray(GameTutorials.president.sectionsJson().map { JSONObject(it as Map<*, *>) }))
            p.put("menuSections", JSONArray(GameTutorials.president.browserMenuSectionsJson().map { JSONObject(it as Map<*, *>) }))
        }
        broadcast(p)
    }

    private fun broadcastSeries() {
        val scores = JSONObject()
        for ((k, v) in series.scores) scores.put(k, v)
        broadcast(JSONObject().put("type", "series_state").put("rounds", series.rounds).put("scores", scores))
    }

    private fun emit() { onStateChange?.invoke() }
    private fun broadcast(o: JSONObject) = server.broadcast(o.toString())
    private fun send(g: GuestId, o: JSONObject) = server.send(g, o.toString())
}

File: ios/Jamboree/Games/President/PresidentServer.swift
import Foundation

/// Wraps [HostServer] with President (Scum/Asshole) routing.
final class PresidentServer {
    private(set) var engine = PresidentEngine()
    static let hostId = "host"
    let hostName: String

    private let server: HostServer
    private var guestToPlayer: [GuestId: String] = [:]
    private var playerToGuest: [String: GuestId] = [:]

    var onStateChange: (() -> Void)?
    private var statRecorded = false
    private let series = SeriesScore()
    // President plays multiple rounds per match (next_round), so the series
    // counts each round's president — gated separately from the once-per-match
    // stat recording.
    private var seriesRecorded = false

    init(server: HostServer? = nil, hostName: String = "Host") {
        self.server = server ?? HostServer(html: HostServer.htmlResource(named: "president_browser"))
        self.hostName = hostName
        self.server.onMessage = { [weak self] id, raw in self?.onMessage(from: id, raw: raw) }
        self.server.onLeave = { [weak self] id in self?.onLeave(id) }
    }

    @discardableResult
    func start() throws -> URL? {
        engine.addPlayer(id: Self.hostId, name: hostName, isHost: true)
        let url = try server.start()
        let local = server.attachLocalGuest()
        guestToPlayer[local] = Self.hostId
        playerToGuest[Self.hostId] = local
        emit()
        return url
    }

    @MainActor
    func makeLoopback() -> LoopbackGuest { LoopbackGuest(server: server) }
    func stop() { server.stop(); resetState() }

    /// Clear all per-session state so the next time the host starts
    /// hosting they get a fresh screen — empty lobby, tutorial vote
    /// available again.
    private func resetState() {
        engine = PresidentEngine()
        guestToPlayer.removeAll(); playerToGuest.removeAll()
        series.reset(); seriesRecorded = false
        statRecorded = false
        emit()
    }
    var guestCount: Int { server.guestCount }

    func hostSetOptions(_ o: PresOptions) {
        engine.setOptions(o); broadcastOptions(); emit()
    }
    func hostStart() {
        seriesRecorded = false
        engine.start()
        broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
    }
    func hostNextRound() {
        seriesRecorded = false
        engine.startNextRound()
        broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
    }
    func hostNewGame() {
        engine.reset(); statRecorded = false
        broadcast(["type": "reset"])
        broadcastLobby(); broadcastTutorialState()
        if !series.isEmpty { broadcastSeries() }
        emit()
    }
    func hostCallTutorialVote() { openTutorialVote() }
    func hostTutorialVote(_ yes: Bool) { submitTutorialVote(voterId: Self.hostId, yes: yes) }
    func hostDismissTutorial() {
        engine.tutorialVote.markShown()
        broadcastTutorialState(); emit()
    }

    private func onMessage(from guest: GuestId, raw: String) {
        guard let data = raw.data(using: .utf8),
              let j = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = j["type"] as? String else { return }
        switch type {
        case "join": handleJoin(guest, json: j)
        case "play":
            if let pid = guestToPlayer[guest] { applyPlay(pid, json: j) }
        case "pass":
            if let pid = guestToPlayer[guest] { applyPass(pid) }
        case "swap":
            if let pid = guestToPlayer[guest] { applySwap(pid, json: j) }
        case "next_round":
            // Anyone can ask, but only honored at gameOver — usually host.
            if engine.phase == .gameOver { hostNextRound() }
        case "set_options":
            break
        case "call_tutorial_vote": if guestToPlayer[guest] != nil { openTutorialVote() }
        case "tutorial_vote":
            if let pid = guestToPlayer[guest], let yes = j["yes"] as? Bool {
                submitTutorialVote(voterId: pid, yes: yes)
            }
        default: break
        }
    }

    private func onLeave(_ guest: GuestId) {
        guard let pid = guestToPlayer.removeValue(forKey: guest) else { return }
        playerToGuest[pid] = nil
        if engine.phase == .lobby {
            engine.removePlayer(pid)
            engine.tutorialVote.removeVoter(pid)
            broadcastLobby()
            if engine.tutorialVote.isOpen || engine.tutorialVote.hasResult {
                broadcastTutorialState()
            }
        }
        emit()
    }

    private func handleJoin(_ guest: GuestId, json: [String: Any]) {
        if engine.phase != .lobby {
            send(guest, ["type": "error", "message": "Game already in progress"])
            return
        }
        let name = String(((json["name"] as? String) ?? "").trimmingCharacters(in: .whitespaces).prefix(24))
        guard !name.isEmpty else { return }
        let pid = "g\(guestToPlayer.count + 1)"
        engine.addPlayer(id: pid, name: name)
        guestToPlayer[guest] = pid
        playerToGuest[pid] = guest
        send(guest, ["type": "welcome", "yourId": pid, "yourName": name, "game": "president"])
        broadcastLobby(); broadcastOptions(); broadcastTutorialState()
        if !series.isEmpty { broadcastSeries() }
        emit()
    }

    private func applyPlay(_ pid: String, json: [String: Any]) {
        guard let cardArr = json["cards"] as? [[String: Any]] else { return }
        var cards: [PresCard] = []
        for c in cardArr {
            guard let s = c["suit"] as? String, let suit = PresSuit(rawValue: s),
                  let r = c["rank"] as? Int else { return }
            cards.append(PresCard(suit: suit, rank: r))
        }
        let err = engine.play(playerId: pid, cards: cards)
        if err == nil {
            broadcastState(); sendHandsPrivately()
            if engine.phase == .gameOver { broadcastOver() }
            emit()
        } else if let g = playerToGuest[pid] {
            send(g, ["type": "error", "message": err!])
        }
    }

    private func applyPass(_ pid: String) {
        let err = engine.pass(playerId: pid)
        if err == nil {
            broadcastState(); emit()
        } else if let g = playerToGuest[pid] {
            send(g, ["type": "error", "message": err!])
        }
    }

    private func applySwap(_ pid: String, json: [String: Any]) {
        guard let cardArr = json["cards"] as? [[String: Any]] else { return }
        var cards: [PresCard] = []
        for c in cardArr {
            guard let s = c["suit"] as? String, let suit = PresSuit(rawValue: s),
                  let r = c["rank"] as? Int else { return }
            cards.append(PresCard(suit: suit, rank: r))
        }
        let err = engine.submitSwap(fromId: pid, cards: cards)
        if err == nil {
            broadcastState(); sendHandsPrivately(); broadcastSwapPrompts(); emit()
        } else if let g = playerToGuest[pid] {
            send(g, ["type": "error", "message": err!])
        }
    }

    private func broadcastOptions() {
        broadcast([
            "type": "options",
            "allowHouseRules": engine.options.allowHouseRules,
            "revolution": engine.options.revolution,
        ])
    }

    private func broadcastLobby() {
        broadcast([
            "type": "lobby",
            "players": engine.players.values.map { ["id": $0.id, "name": $0.name, "isHost": $0.isHost] },
        ])
    }

    private func broadcastState() {
        let phase: String
        switch engine.phase {
        case .lobby: phase = "lobby"
        case .swapping: phase = "swapping"
        case .playing: phase = "playing"
        case .gameOver: phase = "gameOver"
        }
        var payload: [String: Any] = [
            "type": "state",
            "phase": phase,
            "currentId": engine.current?.id as Any,
            "roundNumber": engine.roundNumber,
            "lastEvent": engine.lastEvent ?? "",
            "players": engine.players.values.map {
                [
                    "id": $0.id, "name": $0.name, "handCount": $0.hand.count,
                    "rank": $0.rank.rawValue, "finished": $0.finished,
                    "finishOrder": $0.finishOrder,
                ] as [String: Any]
            },
            "seating": engine.seating,
            "passedThisTrick": Array(engine.passedThisTrick),
        ]
        if let t = engine.trick {
            payload["trick"] = comboJSON(t.combo).merging([
                "topPower": t.topPower,
                "leaderId": t.leaderId,
            ]) { $1 }
        }
        if let lp = engine.lastPlay {
            payload["lastPlay"] = [
                "playerId": lp.playerId,
                "cards": lp.cards.map { ["suit": $0.suit.rawValue, "rank": $0.rank] },
                "combo": comboJSON(lp.combo),
            ]
        }
        broadcast(payload)
    }

    private func comboJSON(_ combo: PresCombo) -> [String: Any] {
        switch combo {
        case .single: return ["kind": "single", "length": 1]
        case .pair: return ["kind": "pair", "length": 2]
        case .triple: return ["kind": "triple", "length": 3]
        case .quad: return ["kind": "quad", "length": 4]
        case .runOfPairs(let len): return ["kind": "runOfPairs", "length": len * 2]
        }
    }

    private func sendHandsPrivately() {
        for p in engine.players.values {
            guard let g = playerToGuest[p.id] else { continue }
            send(g, [
                "type": "hand",
                "cards": p.hand.map { ["suit": $0.suit.rawValue, "rank": $0.rank] },
            ])
        }
    }

    private func broadcastSwapPrompts() {
        // Each player gets a private list of swaps requiring their input.
        for p in engine.players.values {
            guard let g = playerToGuest[p.id] else { continue }
            let mine = engine.pendingSwaps.compactMap { sw -> [String: Any]? in
                guard sw.cards == nil, sw.fromId == p.id else { return nil }
                return [
                    "toId": sw.toId,
                    "toName": engine.players[sw.toId]?.name ?? "",
                    "count": sw.count,
                    "giverChooses": sw.giverChooses,
                ]
            }
            send(g, ["type": "swap_prompts", "prompts": mine])
        }
    }

    private func broadcastOver() {
        if !statRecorded {
            statRecorded = true
            var names: [String] = []
            for pid in engine.finishOrder {
                if let p = engine.players[pid] { names.append(p.name) }
            }
            StatsStore.record(gameId: "president", players: names, outcome: "win")
        }
        if !seriesRecorded {
            seriesRecorded = true
            if let presidentId = engine.finishOrder.first, let p = engine.players[presidentId] {
                series.record(p.name); broadcastSeries()
            }
        }
        let rankings = engine.finishOrder.compactMap { pid -> [String: Any]? in
            guard let p = engine.players[pid] else { return nil }
            return ["id": p.id, "name": p.name, "rank": p.rank.rawValue,
                    "finishOrder": p.finishOrder]
        }
        broadcast([
            "type": "over",
            "rankings": rankings,
        ])
    }

    private func openTutorialVote() {
        guard engine.phase == .lobby, !engine.tutorialVote.isOpen, !engine.tutorialVote.tutorialShown else { return }
        engine.tutorialVote.open(eligibleIds: engine.players.keys)
        broadcastTutorialState(); emit()
    }
    private func submitTutorialVote(voterId: String, yes: Bool) {
        guard engine.tutorialVote.isOpen else { return }
        engine.tutorialVote.submit(voterId: voterId, yes: yes)
        broadcastTutorialState(); emit()
    }
    private func broadcastTutorialState() {
        let v = engine.tutorialVote
        var payload: [String: Any] = [
            "type": "tutorial_vote_state",
            "isOpen": v.isOpen, "yesCount": v.yesCount, "noCount": v.noCount,
            "eligibleCount": v.eligibleCount, "result": v.result as Any, "tutorialShown": v.tutorialShown,
        ]
        if v.result == true && !v.tutorialShown {
            payload["title"] = GameTutorials.president.title
            payload["sections"] = GameTutorials.president.sectionsJSON()
            payload["menuSections"] = GameTutorials.president.browserMenuSectionsJSON()
        }
        broadcast(payload)
    }

    private func broadcastSeries() {
        var scores: [String: Int] = [:]
        for entry in series.scores { scores[entry.key] = entry.value }
        broadcast(["type": "series_state", "rounds": series.rounds, "scores": scores])
    }

    private func emit() { onStateChange?() }
    private func broadcast(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.broadcast(s)
    }
    private func send(_ guest: GuestId, _ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.send(to: guest, s)
    }
}
